# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
# Import your libraries:
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np

# Challenge 1 - Explore the Scikit-Learn Datasets

Before starting to work on our own datasets, let's first explore the datasets that are included in this Python library. These datasets have been cleaned and formatted for use in ML algorithms.

First, we will load the diabetes dataset. Do this in the cell below by importing the datasets and then loading the dataset  to the `diabetes` variable using the `load_diabetes()` function ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_diabetes.html)).

In [2]:
# Your code here:
from sklearn.datasets import load_diabetes
diabetes_dataset=load_diabetes()

Let's explore this variable by looking at the different attributes (keys) of `diabetes`. Note that the `load_diabetes` function does not return dataframes. It returns you a Python dictionary.

In [3]:
# Your code here:
diabetes_dataset.keys()

dict_keys(['data', 'target', 'frame', 'DESCR', 'feature_names', 'data_filename', 'target_filename'])

#### The next step is to read the description of the dataset. 

Print the description in the cell below using the `DESCR` attribute of the `diabetes` variable. Read the data description carefully to fully understand what each column represents.

*Hint: If your output is ill-formatted by displaying linebreaks as `\n`, it means you are not using the `print` function.*

In [4]:
# Your code here:
print(diabetes_dataset["DESCR"])

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - age     age in years
      - sex
      - bmi     body mass index
      - bp      average blood pressure
      - s1      tc, T-Cells (a type of white blood cells)
      - s2      ldl, low-density lipoproteins
      - s3      hdl, high-density lipoproteins
      - s4      tch, thyroid stimulating hormone
      - s5      ltg, lamotrigine
      - s6      glu, blood sugar level

Note: Each of these 10 feature va

#### Based on the data description, answer the following questions:

1. How many attributes are there in the data? What do they mean?

1. What is the relation between `diabetes['data']` and `diabetes['target']`?

1. How many records are there in the data?

Enter your answer here:
1. There are 10 attributes which are the following:
- age: age in years
- sex: sex of the person
- bmi: body mass index
- bp: average blood pressure
- s1: tc, T-Cells (a type of white blood cells)
- s2: ldl, low-density lipoproteins
- s3: hdl, high-density lipoproteins
- s4: tch, thyroid stimulating hormone
- s5: ltg, lamotrigine
- s6: glu, blood sugar level
    
All of them are standarized, they are rescaled to have a mean equal to 0 and a standard deviation equal to 1, for this reason, for example, the age can be negative.
    
2. The data are the variables used to predict the target

3. There are 442 records

#### Now explore what are contained in the *data* portion as well as the *target* portion of `diabetes`. 

Scikit-learn typically takes in 2D numpy arrays as input (though pandas dataframes are also accepted). Inspect the shape of `data` and `target`. Confirm they are consistent with the data description.

In [5]:
# Your code here:
diabetes_data=pd.DataFrame(diabetes_dataset["data"], columns=diabetes_dataset["feature_names"])
diabetes_data.shape

(442, 10)

In [7]:
diabetes_target=pd.DataFrame(diabetes_dataset["target"])
diabetes_target.shape

(442, 1)

Both have 442 records

# Challenge 2 - Perform Supervised Learning on the Dataset

The data have already been split to predictor (*data*) and response (*target*) variables. Given this information, we'll apply what we have previously learned about linear regression and apply the algorithm to the diabetes dataset.

#### Let's briefly revisit the linear regression formula:

```
y = β0 + β1X1 + β2X2 + ... + βnXn + ϵ
```

...where:

- X1-Xn: data 
- β0: intercept 
- β1-βn: coefficients 
- ϵ: error (cannot explained by model)
- y: target

Also take a look at the `sklearn.linear_model.LinearRegression` [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html).

#### In the cell below, import the `linear_model` class from `sklearn`. 

In [8]:
# Your code here:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

#### Create a new instance of the linear regression model and assign the new instance to the variable `diabetes_model`.

In [9]:
# Your code here:
diabetes_data["diabetes_model"]=diabetes_dataset["target"]
diabetes_data.head()

age       sex       bmi        bp        s1        s2        s3  \
0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2  0.085299  0.050680  0.044451 -0.005671 -0.045599 -0.034194 -0.032356   
3 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   

         s4        s5        s6  diabetes_model  
0 -0.002592  0.019908 -0.017646           151.0  
1 -0.039493 -0.068330 -0.092204            75.0  
2 -0.002592  0.002864 -0.025930           141.0  
3  0.034309  0.022692 -0.009362           206.0  
4 -0.002592 -0.031991 -0.046641           135.0

#### Next, let's split the training and test data.

Define `diabetes_data_train`, `diabetes_target_train`, `diabetes_data_test`, and `diabetes_target_test`. Use the last 20 records for the test data and the rest for the training data.

In [13]:
# Your code here:
y=diabetes_data["diabetes_model"]
X=diabetes_data.drop(columns="diabetes_model")
diabetes_data_train, diabetes_data_test, diabetes_target_train, diabetes_target_test=train_test_split(X,y, test_size=0.20)

Fit the training data and target to `diabetes_model`. Print the *intercept* and *coefficients* of the model.

In [14]:
# Your code here:
lr=LinearRegression()
lr.fit(diabetes_data_train, diabetes_target_train)

LinearRegression()

In [15]:
lr.intercept_

152.61732796671168

In [16]:
lr.coef_

array([ -40.0333523 , -195.85927611,  504.00837366,  324.29776851,
       -813.40305768,  447.79561464,  108.09323763,  260.48739259,
        728.92682505,   83.64455817])

#### Inspecting the results

From the outputs you should have seen:

- The intercept is a float number.
- The coefficients are an array containing 10 float numbers.

This is the linear regression model fitted to your training dataset.

#### Using your fitted linear regression model, predict the *y* of `diabetes_data_test`.

In [25]:
# Your code here:
y_pred = lr.predict(diabetes_data_test)
array1=np.array(y_pred)

#### Print your `diabetes_target_test` and compare with the prediction. 

In [26]:
# Your code here:
array2=np.array(diabetes_target_test)

In [27]:
array1==array2

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False])

#### Is `diabetes_target_test` exactly the same as the model prediction? Explain.

In [ ]:
# Your explanation here:
# They are not the same as the values are a prediction

# Bonus Challenge 1 - Hypothesis Testing with `statsmodels`

After generating the linear regression model from the dataset, you probably wonder: then what? What is the statistical way to know if my model is reliable or not?

Good question. We'll discuss that using Scikit-Learn in Challenge 5. But for now, let's use a fool-proof way by using the ([Linear Regression class of StatsModels](https://www.statsmodels.org/dev/regression.html)) which can also conduct linear regression analysis plus much more such as calcuating the F-score of the linear model as well as the standard errors and t-scores for each coefficient. The F-score and t-scores will tell you whether you can trust your linear model.

To understand the statistical meaning of conducting hypothesis testing (e.g. F-test, t-test) for slopes, read [this webpage](https://onlinecourses.science.psu.edu/stat501/node/297/) at your leisure time. We'll give you a brief overview next.

* The F-test of your linear model is to verify whether at least one of your coefficients is significantly different from zero. Translating that into the *null hypothesis* and *alternative hypothesis*, that is:

    ```
    H0 : β1 = β2 = ... = β10 = 0
    HA : At least one βj ≠ 0 (for j = 1, 2, ..., 10)
    ```

* The t-tests on each coefficient is to check whether the confidence interval for the variable contains zero. If the confidence interval contains zero, it means the null hypothesis for that variable is not rejected. In other words, this particular vaiable is not contributing to your linear model and you can remove it from your formula.

Read the documentations of [StatsModels Linear Regression](https://www.statsmodels.org/dev/regression.html) as well as its [`OLS` class](https://www.statsmodels.org/dev/generated/statsmodels.regression.linear_model.OLS.html) which stands for *ordinary least squares*.

#### In the next cell, analyze `diabetes_data_train` and `diabetes_target_train` with the linear regression model of `statsmodels`. Print the fit summary.

Your output should look like:

![statsmodels regression](../statsmodels.png)

In [57]:
# Your code here:
import statsmodels.api as smf


In [62]:
mod = smf.OLS(diabetes_target_train, diabetes_data_train)

In [63]:
res = mod.fit()

In [64]:
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:         diabetes_model   R-squared (uncentered):                   0.116
Model:                            OLS   Adj. R-squared (uncentered):              0.090
Method:                 Least Squares   F-statistic:                              4.481
Date:                Thu, 03 Dec 2020   Prob (F-statistic):                    5.86e-06
Time:                        00:11:50   Log-Likelihood:                         -2295.5
No. Observations:                 353   AIC:                                      4611.
Df Residuals:                     343   BIC:                                      4650.
Df Model:                          10                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
age         -125.6984    201.271     -0.625      0.533    -521.580     270.183
sex         -228.7347    205.338     -1.114      0.266    -632.615     175.145
bmi          725.0267    222.806      3.254      0.001     286.789    1163.264
bp           302.6198    217.869      1.389      0.166    -125.907     731.146
s1         -1101.2831   1542.155     -0.714      0.476   -4134.554    1931.987
s2           817.7756   1284.379      0.637      0.525   -1708.475    3344.026
s3           324.7072    740.442      0.439      0.661   -1131.671    1781.086
s4           321.7494    550.862      0.584      0.560    -761.743    1405.242
s5           656.7077    615.121      1.068      0.286    -553.176    1866.591
s6            67.1469    226.616      0.296      0.767    -378.585     512.879
==============================================================================
Omnibus:                        1.498   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.473   Jarque-Bera (JB):                1.442
Skew:                          -0.065   Prob(JB):                        0.486
Kurtosis:                       2.715   Cond. No.                         23.8
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### Interpreting hypothesis testing results

Answer the following questions in the cell below:

1. What is the F-score of your linear model and is the null hypothesis rejected?

1. Does any of the t-tests of the coefficients produce a confidence interval containing zero? What are they?

1. How will you modify your linear reguression model according to the test results above?


1. The F score is 4.481, the null hypothesis is rejected

# Challenge 3 - Peform Supervised Learning on a Pandas Dataframe

Now that we have dealt with data that has been formatted for scikit-learn, let's look at data that we will need to format ourselves.

In the next cell, load the `auto-mpg.csv` file included in this folder and assign it to a variable called `auto`.

In [28]:
# Your code here:
auto = pd.read_csv('../auto-mpg.csv')


Look at the first 5 rows using the `head()` function:

In [29]:
# Your code here:
auto.head()


mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

Evaluate the data to ensure that all numeric columns are correctly detected as such by pandas. If a column is misclassified as object, coerce it to numeric.

In [30]:
# Your code here:
auto.dtypes


mpg             float64
cylinders         int64
displacement    float64
horse_power     float64
weight            int64
acceleration    float64
model_year        int64
car_name         object
dtype: object

In [ ]:
auto["car_name"] = auto["car_name"].astype(str).astype(int)

What is the newest model year and the oldest model year?

In [33]:
# Your code here:
newest=auto["model_year"].max()
oldest=auto["model_year"].min()
newest, oldest

(82, 70)

Check the dataset for missing values and remove all rows containing at least one missing value.

In [34]:
# Your code here:
auto.isnull().sum()


mpg             0
cylinders       0
displacement    0
horse_power     6
weight          0
acceleration    0
model_year      0
car_name        0
dtype: int64

In [35]:
auto.dropna(axis=0, inplace=True)
auto.head()

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

Find the frequency table for the `cylinders` column using the `value_counts()` function. How many possible values of cylinders are there?

In [36]:
# Your code here:
auto["cylinders"].value_counts()

4    199
8    103
6     83
3      4
5      3
Name: cylinders, dtype: int64

We would like to generate a linear regression model that will predict mpg. To do this, first drop the `car_name` column since it does not contain any quantitative data. Next separate the dataframe to predictor and response variables. Separate those into test and training data with 80% of the data in the training set and the remainder in the test set. 

Assign the predictor and response training data to `X_train` and `y_train` respectively. Similarly, assign the predictor and response test data to `X_test` and `y_test`.

*Hint: To separate data for training and test, use the `train_test_split` method we used in previous labs.*

In [37]:
# Your code here:
auto.drop(columns="car_name", inplace=True)


In [38]:
y=auto["mpg"]
X=auto.drop(columns="mpg")
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20)

Now we will processed and peform linear regression on this data to predict the mpg for each vehicle. 

#### In the next cell, create an instance of the linear regression model and call it `auto_model`. Fit `auto_model` with your training data.

In [40]:
auto_model=LinearRegression()
auto_model.fit(X_train, y_train)

LinearRegression()

# Challenge 4 - Evaluate the Model

In addition to evaluating your model with F-test and t-test, you can also use the *Coefficient of Determination* (a.k.a. *r squared score*). This method does not simply tell *yes* or *no* about the model fit but instead indicates how much variation can be explained by the model. Based on the r squared score, you can decide whether to improve your model in order to obtain a better fit.

You can learn about the r squared score [here](). Its formula is:

![R Squared](../r-squared.png)

...where:

* yi is an actual data point.
* ŷi is the corresponding data point on the estimated regression line.

By adding the squares of the difference between all yi-ŷi pairs, we have a measure called SSE (*error sum of squares*) which is an application of the r squared score to indicate the extent to which the estimated regression model is different from the actual data. And we attribute that difference to the random error that is unavoidable in the real world. Obviously, we want the SSE value to be as small as possible.

#### In the next cell, compute the predicted *y* based on `X_train` and call it `y_pred`. Then calcualte the r squared score between `y_pred` and `y_train` which indicates how well the estimated regression model fits the training data.

*Hint: r squared score can be calculated using `sklearn.metrics.r2_score` ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html)).*

In [41]:
# Your code here:
from sklearn.metrics import r2_score


In [43]:
y_pred=auto_model.predict(X_train)
r2_score(y_train,y_pred)

0.8156304323414025

#### Our next step is to evaluate the model using the test data. 

We would like to ensure that our model is not overfitting the data. This means that our model was made to fit too closely to the training data by being overly complex. If a model is overfitted, it is not generalizable to data outside the training data. In that case, we need to reduce the complexity of the model by removing certain features (variables).

In the cell below, use the model to generate the predicted values for the test data and assign them to `y_test_pred`. Compute the r squared score of the predicted `y_test_pred` and the oberserved `y_test` data.

In [47]:
# Your code here:
y_test_pred=auto_model.predict(X_test)


In [49]:
r2_score(y_test, y_test_pred)

0.780054452159149

#### Explaining the results

The r squared scores of the training data and the test data are pretty close (0.8146 vs 0.7818). This means our model is not overfitted. However, there is still room to improve the model fit. Move on to the next challenge.

# Challenge 5 - Improve the Model Fit

While the most common way to improve the fit of a model is by using [regularization](https://datanice.github.io/machine-learning-101-what-is-regularization-interactive.html), there are other simpler ways to improve model fit. The first is to create a simpler model. The second is to increase the train sample size.

Let us start with the easier option and increase our train sample size to 90% of the data. Create a new test train split and name the new predictors and response variables `X_train09`, `X_test09`, `y_train09`, `y_test09`.

In [50]:
# Your code here:
y=auto["mpg"]
X=auto.drop(columns="mpg")
X_train09, X_test09, y_train09, y_test09 = train_test_split(X, y, test_size=0.10)

Initialize a new linear regression model. Name this model `auto_model09`. Fit the model to the new sample (training) data.

In [51]:
# Your code here:
auto_model09=LinearRegression()
auto_model09.fit(X_train09, y_train09)

LinearRegression()

Compute the predicted values and r squared score for our new model and new sample data.

In [52]:
# Your code here:
auto_model09.intercept_


-15.217166285919216

In [53]:
coeficients=auto_model09.coef_
coeficients

array([-0.26788241,  0.00942141, -0.00482999, -0.00685744,  0.1484327 ,
        0.75030004])

In [54]:
y_pred09 = auto_model09.predict(X_train09)
r2_score(y_train09,y_pred09)

0.8110992027511301

Compute the r squared score for the smaller test set. Is there an improvement in the test r squared?

In [55]:
# Your code here:
y_pred09_test = auto_model09.predict(X_test09)
r2_score(y_test09,y_pred09_test)

0.7591910021484699

The r2 score of the test data set in smaller so the line is worst fitted to the points

# Bonus Challenge 2 - Backward Elimination 

The main way to produce a simpler linear regression model is to reduce the number of variables used in the model. In scikit-learn, we can do this by using recursive feature elimination. You can read more about RFE [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html).

In the next cell, we will import RFE

In [65]:
from sklearn.feature_selection import RFE

Follow the documentation and initialize an RFE model using the `auto_model` linear regression model. Set `n_features_to_select=3`

In [66]:
# Your code here:
selector = RFE(coeficients, n_features_to_select=3)


Fit the model and print the ranking

In [67]:
# Your code here:
selector = selector.fit(X_train, y_train)

AttributeError: 'numpy.ndarray' object has no attribute '_get_tags'

Feature importance is ranked from most important (1) to least important (4). Generate a model with the three most important features. The features correspond to variable names. For example, feature 1 is `cylinders` and feature 2 is `displacement`.

Perform a test-train split on this reduced column data and call the split data `X_train_reduced`, `X_test_reduced`, `y_test_reduced`, `y_train_reduced`. Use an 80% split.

In [ ]:
# Your code here:


Generate a new model called `auto_model_reduced` and fit this model. Then proceed to compute the r squared score for the model. Did this cause an improvement in the r squared score?

In [ ]:
# Your code here: 


#### Conclusion

You may obtain the impression from this lab that without knowing statistical methods in depth, it is difficult to make major progress in machine learning. That is correct. If you are motivated to become a data scientist, statistics is the subject you must be proficient in and there is no shortcut. 

Completing these labs is not likely to make you a data scientist. But you will have a good sense about what are there in machine learning and what are good for you. In your future career, you can choose one of the three tracks:

* Data scientists who need to be proficient in statistical methods.

* Data engineers who need to be good at programming.

* Data integration specialists who are business or content experts but also understand data and programming. This cross-disciplinary track brings together data, technology, and business and will be in high demands in the next decade.